### PR TITLE
allow to find actor by ain & reload

### PR DESCRIPTION
**Pull request by klausmeyer**
See #34

### CODE REVIEW

File: lib/fritzbox/smarthome/actor.rb
Comment (carpodaster) [inline lib/fritzbox/smarthome/actor.rb]: We can use Ruby mandatory keywords for this:

```suggestion
        def find_by!(ain:)
```

File: lib/fritzbox/smarthome/actor.rb
Comment (carpodaster) [inline lib/fritzbox/smarthome/actor.rb]:
```suggestion
          raise ArgumentError, 'Find via :id is not implemented yet' if id.present?
```

But what would an id be? I would think the AIN already is a unique identifier

File: lib/fritzbox/smarthome/actor.rb
Comment (carpodaster) [inline lib/fritzbox/smarthome/actor.rb:61]: I wonder if all `StandardError`s should be rescued. Could also be a parser error, or some timeouts raised by the network call

File: lib/fritzbox/smarthome/actor.rb
Comment (carpodaster) [inline lib/fritzbox/smarthome/actor.rb]: :question:  **Optional**

Would also return the `ain` of the resource:

```suggestion
        raise ResourceNotFound.new("Unable to reload self with ain #{ain.inspect}")
```

File: examples/refresh.rb
Comment (carpodaster) [inline examples/refresh.rb]: To illustrate the use case that one doesn't have to fetch the full list first:

```suggestion
pp actor = Fritzbox::Smarthome::Actor.find_by!(ain: "0815-4711") # example AIN
```

File: lib/fritzbox/smarthome/actor.rb
Comment (carpodaster) [inline lib/fritzbox/smarthome/actor.rb:57]: :thinking:  Maybe call it `reload!` because of the network call side-effect and changing the state of `self`?

**Review by carpodaster — COMMENTED** [review]
:confetti_ball: 

Added some suggestions. 

I'm not at a machine where I can test it right now, but I'd be interested in benchmarks. My API responses are always comparatively slow. I wonder if only returning one record on the Fritz!Box end makes it significantly faster.

File: lib/fritzbox/smarthome/actor.rb
Comment (klausmeyer) [reply lib/fritzbox/smarthome/actor.rb]: True. I was already thinking ahead of cases where someone would like to fetch with the `:id` instead of the `:ain` - but I guess it's better to keep it for the future.

File: lib/fritzbox/smarthome/actor.rb
Comment (klausmeyer) [reply lib/fritzbox/smarthome/actor.rb]: This one here:

https://github.com/klausmeyer/fritzbox-smarthome/blob/aca8da316bc48f9442f5121fa0d47258ee439ec3/lib/fritzbox/smarthome/actor.rb#L27

It's returned as a numeric ID whereas the AIN looks like something based on the serial number or hardware address. The ID seems to be pretty useless as it can't be used for other operations.

Thanks for pointing me to the typo - but as mentioned above I guess I'll remove that part completely to simplify.

File: lib/fritzbox/smarthome/actor.rb
Comment (klausmeyer) [reply lib/fritzbox/smarthome/actor.rb:61]: In this case I'm using the KeyError to detect real "404" reponses where the body is returned as '{}'.

Regular StandardErrors I would prefer to to keep bubbling up as they might be more informative then having them hidden behind a `ResourceNotFound` which wouldn't fit for all cases.

File: lib/fritzbox/smarthome/actor.rb
Comment (klausmeyer) [reply lib/fritzbox/smarthome/actor.rb]: Makes sense.

File: examples/refresh.rb
Comment (klausmeyer) [reply examples/refresh.rb]: Yes that would make sense.

I'll check how I will do that and still keep the example run-able for everyone.

Intention here was to not hardcode an AIN as not everyone would have this corresponding device on their FritzBox ;)

File: lib/fritzbox/smarthome/actor.rb
Comment (klausmeyer) [reply lib/fritzbox/smarthome/actor.rb:57]: I'm a bit trying to keep the interface similar to `ActiveRecord` and there it's also just called `reload` (without the bang).

**Comment by klausmeyer** [comment]
@carpodaster Thank you for 👀

> I'm not at a machine where I can test it right now, but I'd be interested in benchmarks. My API responses are always comparatively slow. I wonder if only returning one record on the Fritz!Box end makes it significantly faster.

Would be interesting indeed.

In general the APIs are quite slow for me as well.

**Review by carpodaster — APPROVED** [review]
:shipit:

**Comment by carpodaster** [comment]
.oO0{ why can I review this PR in the first place? }  :thinking:  :smile:

**Comment by klausmeyer** [comment]
Guess in theory everyone can.

Might be some weird default in GitHub - but I actually like it.

I'll wait for the feedback from @mkoegel in #34 before merging.

Also need to put a few specs around the new stuff I guess ;)

**Pull request merged**
